### ISSUE TITLE
Breaking change in Python 3.10rc1

### ISSUE THREAD

**Issue by frenzymadness** [ipython/ipython#13079]
Hello.

There seems to be one more incompatibility in Python 3.10rc1. The problem is in this function: https://github.com/ipython/ipython/blob/770024afeaa5bd66910cf89b9667c4967b65004e/IPython/core/oinspect.py#L287-L319

The call of `inspect.getabsfile(obj.__class__)` on line 313 raises an OSError instead of the expected TypeError. The change in the CPython is here: https://github.com/python/cpython/pull/27495/files

Small reproducer derived from `test_debugger.test_ipdb_magics` is:
```
Python 3.10.0rc1 (default, Aug  3 2021, 00:00:00) [GCC 11.2.1 20210728 (Red Hat 11.2.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from IPython.core import debugger
>>> class ExampleClass(object):
...     """Docstring for ExampleClass."""
...     def __init__(self):
...         """Docstring for ExampleClass.__init__"""
...         pass
...     def __str__(self):
...         return "ExampleClass()"
... 
>>> a = ExampleClass()
>>> debugger.Pdb().set_trace()
--Return--
None
> <stdin>(1)<module>()

ipdb> pinfo a
Traceback (most recent call last):
  File "/home/lbalhar/Software/ipython/IPython/core/oinspect.py", line 307, in find_file
    fname = inspect.getabsfile(obj)
  File "/usr/lib64/python3.10/inspect.py", line 844, in getabsfile
    _filename = getsourcefile(object) or getfile(object)
  File "/usr/lib64/python3.10/inspect.py", line 817, in getsourcefile
    filename = getfile(object)
  File "/usr/lib64/python3.10/inspect.py", line 797, in getfile
    raise TypeError('module, class, method, function, traceback, frame, or '
TypeError: module, class, method, function, traceback, frame, or code object was expected, got ExampleClass

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.10/bdb.py", line 94, in trace_dispatch
    return self.dispatch_return(frame, arg)
  File "/usr/lib64/python3.10/bdb.py", line 153, in dispatch_return
    self.user_return(frame, arg)
  File "/usr/lib64/python3.10/pdb.py", line 294, in user_return
    self.interaction(frame, None)
  File "/home/lbalhar/Software/ipython/IPython/core/debugger.py", line 356, in interaction
    OldPdb.interaction(self, frame, traceback)
  File "/usr/lib64/python3.10/pdb.py", line 357, in interaction
    self._cmdloop()
  File "/usr/lib64/python3.10/pdb.py", line 322, in _cmdloop
    self.cmdloop()
  File "/usr/lib64/python3.10/cmd.py", line 138, in cmdloop
    stop = self.onecmd(line)
  File "/usr/lib64/python3.10/pdb.py", line 422, in onecmd
    return cmd.Cmd.onecmd(self, line)
  File "/usr/lib64/python3.10/cmd.py", line 217, in onecmd
    return func(arg)
  File "/home/lbalhar/Software/ipython/IPython/core/debugger.py", line 778, in do_pinfo
    self.shell.find_line_magic("pinfo")(arg, namespaces=namespaces)
  File "/home/lbalhar/.virtualenvs/ipython/lib/python3.10/site-packages/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)
  File "/home/lbalhar/Software/ipython/IPython/core/magic.py", line 187, in <lambda>
    call = lambda f, *a, **k: f(*a, **k)
  File "/home/lbalhar/Software/ipython/IPython/core/magics/namespace.py", line 58, in pinfo
    self.shell._inspect('pinfo', oname, detail_level=detail_level,
  File "/home/lbalhar/Software/ipython/IPython/core/interactiveshell.py", line 1779, in _inspect
    pmethod(
  File "/home/lbalhar/Software/ipython/IPython/core/oinspect.py", line 680, in pinfo
    info = self._get_info(obj, oname, formatter, info, detail_level)
  File "/home/lbalhar/Software/ipython/IPython/core/oinspect.py", line 586, in _get_info
    info = self._info(obj, oname=oname, info=info, detail_level=detail_level)
  File "/home/lbalhar/Software/ipython/IPython/core/oinspect.py", line 785, in _info
    fname = find_file(obj)
  File "/home/lbalhar/Software/ipython/IPython/core/oinspect.py", line 313, in find_file
    fname = inspect.getabsfile(obj.__class__)
  File "/usr/lib64/python3.10/inspect.py", line 844, in getabsfile
    _filename = getsourcefile(object) or getfile(object)
  File "/usr/lib64/python3.10/inspect.py", line 817, in getsourcefile
    filename = getfile(object)
  File "/usr/lib64/python3.10/inspect.py", line 785, in getfile
    raise OSError('source code not available')
OSError: source code not available
```

I can confirm that the reproducer does not raise the error with Python 3.10b4. An easy fix might be to expect also OSError from the mentioned function. This workaround seems to work for me so I'm gonna open a PR with it.

**Comment by frenzymadness** [comment]
I'm already working on it. There are two more currently failing tests in test_completer:
```
__________________________ TestCompleter.test_all_completions_dups ___________________________

self = <IPython.core.tests.test_completer.TestCompleter testMethod=test_all_completions_dups>

    def test_all_completions_dups(self):
        """
        Make sure the output of `IPCompleter.all_completions` does not have
        duplicated prefixes.
        """
        ip = get_ipython()
        c = ip.Completer
        ip.ex("class TestClass():\n\ta=1\n\ta1=2")
        for jedi_status in [True, False]:
            with provisionalcompleter():
                ip.Completer.use_jedi = jedi_status
                matches = c.all_completions("TestCl")
                assert matches == ['TestClass'], jedi_status
                matches = c.all_completions("TestClass.")
>               assert len(matches) > 2, jedi_status
E               AssertionError: True
E               assert 0 > 2
E                +  where 0 = len([])

IPython/core/tests/test_completer.py:450: AssertionError
_________________________ TestCompleter.test_deduplicate_completions _________________________

self = <IPython.core.tests.test_completer.TestCompleter testMethod=test_deduplicate_completions>

    def test_deduplicate_completions(self):
        """
        Test that completions are correctly deduplicated (even if ranges are not the same)
        """
        ip = get_ipython()
        ip.ex(
            textwrap.dedent(
                """
        class Z:
            zoo = 1
        """
            )
        )
        with provisionalcompleter():
            ip.Completer.use_jedi = True
            l = list(
                _deduplicate_completions("Z.z", ip.Completer.completions("Z.z", 3))
            )
            ip.Completer.use_jedi = False
    
>       assert len(l) == 1, "Completions (Z.z<tab>) correctly deduplicate: %s " % l
E       AssertionError: Completions (Z.z<tab>) correctly deduplicate: [] 
E       assert 0 == 1
E        +  where 0 = len([])

IPython/core/tests/test_completer.py:524: AssertionError
```
And test_magic_arguments which is already fixed by #12759 . If I manage to fix the new failures, I'll combine these two PRs to one.

**Comment by frenzymadness** [comment]
If I enable the debug mode for the completer both tests fail with something like:
```
E       AssertionError: assert 'Oops Jedi ha...ailable\ns"""' == 'zoo'
E         - zoo
E         + Oops Jedi has crashed, please report a bug with the following:
E         + """
E         + source code not available
E         + s"""
```
Which is exactly the error the updated inspect module newly produces. I'm gonna take a look to the jedi's source code.
